File: C#/score4.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace score4
{
    class Program
    {
        public const int width = 7;
        public const int height = 6;
        public const int maxDepth = 7;
        public const int orangeWins = 1000000;
        public const int yellowWins = -orangeWins;

        public enum Mycell {
            Orange=1,
            Yellow=-1,
            Barren=0
        };

        public class Board {
            // Initially, this was Mycell[,]
            // Unfortunately, C# 2D arrays are a lot slower
            // than simple arrays of arrays (Jagged arrays): Mycell[][]
	    // BUT
	    // using a 1D array is EVEN faster:
	    //    _slots[width*Y + X]
	    // is much faster than
	    //    _slots[Y][X]
	    //
	    // (sigh) Oh well, C# is a VM-based language (specifically, .NET).
	    // Running fast is not the primary concern in VMs...
            public Mycell[] _slots;
            public Board()
            {
               _slots = new Mycell[height*width];
            }
        };

        public static int ScoreBoard(Board board)
        {
            int[] counters = { 0, 0, 0, 0, 0, 0, 0, 0, 0 };

            // Horizontal spans
            for (int y = 0; y < height; y++)
            {
                int score = (int) board._slots[width*(y)+0] + (int) board._slots[width*(y)+1] + (int) board._slots[width*(y)+2];
                for (int x = 3; x < width; x++)
                {
                    score += (int) board._slots[width*(y)+x];
                    counters[score + 4]++;
                    score -= (int) board._slots[width*(y)+x - 3];
                }
            }
            // Vertical spans
            for (int x = 0; x < width; x++)
            {
                int score = (int) board._slots[width*(0)+x] + (int) board._slots[width*(1)+x] + (int) board._slots[width*(2)+x];
                for (int y = 3; y < height; y++)
                {
                    score += (int) board._slots[width*(y)+x];
                    counters[score + 4]++;
                    score -= (int) board._slots[width*(y - 3)+x];
                }
            }
            // Down-right (and up-left) diagonals
            for (int y = 0; y < height - 3; y++)
            {
                for (int x = 0; x < width - 3; x++)
                {
                    int score = 0;
                    for (int ofs=0; ofs<4; ofs++)
                    {
                        int yy = y + ofs;
                        int xx = x + ofs;
                        score += (int) board._slots[width*(yy)+xx];
                    }
                    counters[score + 4]++;
                }
            }
            // up-right (and down-left) diagonals
            for (int y = 3; y < height; y++)
            {
                for (int x = 0; x < width - 3; x++)
                {
                    int score = 0;
                    for (int ofs=0; ofs<4; ofs++)
                    {
                        int yy = y - ofs;
                        int xx = x + ofs;
                        score += (int) board._slots[width*(yy)+xx];
                    }
                    counters[score + 4]++;
                }
            }

            if (counters[0] != 0)
                return yellowWins;
            else if (counters[8] != 0)
                return orangeWins;
            else
                return
                    counters[5] + 2 * counters[6] + 5 * counters[7] + 10 * counters[8] -
                    counters[3] - 2 * counters[2] - 5 * counters[1] - 10 * counters[0];
        }

        public static int dropDisk(Board board, int column, Mycell color)
        {
            for (int y=height-1; y>=0; y--)
                if (board._slots[width*(y)+column] == Mycell.Barren) {
                    board._slots[width*(y)+column] = color;
                    return y;
            }
            return -1;
        }

        public static bool g_debug = false;

        public static Board loadBoard(string[] args)
        {
            Board newBoard = new Board();
            foreach(var arg in args)
                if (arg[0] == 'o' || arg[0] == 'y')
                    newBoard._slots[width*(arg[1]-'0')+arg[2]-'0'] = (arg[0] == 'o')?Mycell.Orange:Mycell.Yellow;
            else if (arg == "-debug")
                g_debug = true;
            return newBoard;
        }

        public static void abMinimax(bool maximizeOrMinimize, Mycell color, int depth, Board board, out int move, out int score)
        {
            if (0 == depth) {
                move = -1;
                score = ScoreBoard(board);
            } else {
                int bestScore=maximizeOrMinimize?-10000000:10000000;
                int bestMove=-1;
                for (int column=0; column<width; column++) {
                    if (board._slots[width*(0)+column]!=Mycell.Barren)
                        continue;
                    int rowFilled = dropDisk(board, column, color);
                    if (rowFilled == -1)
                        continue;
                    int s = ScoreBoard(board);
                    if (s == (maximizeOrMinimize?orangeWins:yellowWins)) {
                        bestMove = column;
                        bestScore = s;
                        board._slots[width*(rowFilled)+column] = Mycell.Barren;
                        break;
                    }
                    int moveInner, scoreInner;
		    if (depth>1)
			abMinimax(!maximizeOrMinimize, color==Mycell.Orange?Mycell.Yellow:Mycell.Orange, depth-1, board, out moveInner, out scoreInner);
		    else {
			moveInner = -1;
			scoreInner = s;
		    }
                    board._slots[width*(rowFilled)+column] = Mycell.Barren;
                    /* when loss is certain, avoid forfeiting the match, by shifting scores by depth... */
                    if (scoreInner == orangeWins || scoreInner == yellowWins)
                        scoreInner -= depth * (int)color;
                    if (depth == maxDepth && g_debug)
                        Console.WriteLine("Depth {0}, placing on {1}, score:{2}", depth, column, scoreInner);
                    if (maximizeOrMinimize) {
                        if (scoreInner>=bestScore) {
                            bestScore = scoreInner;
                            bestMove = column;
                        }
                    } else {
                        if (scoreInner<=bestScore) {
                            bestScore = scoreInner;
                            bestMove = column;
                        }
                    }
                }
                move = bestMove;
                score = bestScore;
            }
        }


        public static void Main(string[] args)
        {
            Board board = loadBoard(args);
            int scoreOrig = ScoreBoard(board);
            if (scoreOrig == orangeWins) { Console.WriteLine("I win"); Environment.Exit(-1); }
            else if (scoreOrig == yellowWins) { Console.WriteLine("You win"); Environment.Exit(-1); }
            else {
                int move, score;
                abMinimax(true, Mycell.Orange, maxDepth, board, out move, out score);
                if (move != -1) {
                    Console.WriteLine("{0}", move);
                    dropDisk(board, move, Mycell.Orange);
                    scoreOrig = ScoreBoard(board);
                    if (scoreOrig == orangeWins) { Console.WriteLine("I win"); Environment.Exit(-1); }
                    else if (scoreOrig == yellowWins) { Console.WriteLine("You win\n"); Environment.Exit(-1); }
                    else Environment.Exit(0);
                } else {
                    Console.WriteLine("No move possible");
                    Environment.Exit(-1);
                }
            }
            Environment.Exit(0);
        }
    }
}

File: Java/Score4.java
import java.util.ArrayList;

public class Score4 {

    static final int width = 7;
    static final int height = 6;
    static final int orangeWins = 1000000;
    static final int yellowWins = -orangeWins;
    static int g_maxDepth = 7;
    static boolean g_debug = false;

    static class Board {

        final int width;
        final int height;
        int[][] slots;

        Board(int width, int height) {
            slots = new int[height][width];
            this.width = width;
            this.height = height;
            fill(0);
        }

        private void fill(int type) {
            for (int j = 0; j < this.height; j++) {
                for (int i = 0; i < this.width; i++) {
                    slots[j][i] = type;
                }
            }
        }

        boolean inside(int y, int x) {
            return y >= 0 && y < this.height && x >= 0 && x < this.width;
        }

        int dropDisk(int column, int color) {
            for (int y = this.height - 1; y >= 0; y--) {
                if (slots[y][column] == 0) {
                    slots[y][column] = color;
                    return y;
                }
            }
            return -1;
        }

        void load(Integer[] x, Integer[] y, Integer[] color) {
            assert ((x.length == y.length) && (x.length == color.length));
            for (int i = 0; i < x.length; i++) {
                slots[y[i]][x[i]] = color[i];
            }
        }
    }

    public static void main(String[] args) {
        Board board = prepareBoard(args);

        int scoreOrig = scoreBoard(board);
        if (scoreOrig == orangeWins) {
            System.out.println("I win");
            System.exit(-1);
        } else if (scoreOrig == yellowWins) {
            System.out.println("You win");
            System.exit(-1);
        } else {
            ScoreMoveTuple smt = abMinimax(true, 1, g_maxDepth, board);
            if (smt.move != -1) {
                System.out.println(smt.move);
                board.dropDisk(smt.move, 1);
                scoreOrig = scoreBoard(board);
                if (scoreOrig == orangeWins) {
                    System.out.println("I win");
                    System.exit(-1);
                } else if (scoreOrig == yellowWins) {
                    System.out.println("You win");
                    System.exit(-1);
                } else {
                    System.exit(0);
                }
            } else {
                System.out.println("No move possible!");
                System.exit(-1);
            }
        }
    }

    private static Board prepareBoard(String[] args) {
        Board board = new Board(width, height);
        ArrayList<Integer> colors = new ArrayList<Integer>(9);
        ArrayList<Integer> x = new ArrayList<Integer>(9);
        ArrayList<Integer> y = new ArrayList<Integer>(9);
        Integer i = 0;
        for (String arg : args) {
            char[] cargs = arg.toCharArray();
            if (arg.length() == 3) {
                if (cargs[0] == 'o') {
                    colors.add(1);
                    y.add(cargs[1] - '0');
                    x.add(cargs[2] - '0');
                } else if (cargs[0] == 'y') {
                    colors.add(-1);
                    y.add(cargs[1] - '0');
                    x.add(cargs[2] - '0');
                }
            } else if (arg.equalsIgnoreCase("-debug")) {
                g_debug = true;
            } else if (arg.equalsIgnoreCase("-level")) {
                g_maxDepth = Integer.getInteger(args[i + 1]);
            }
            i++;
        }
        board.load(x.toArray(new Integer[x.size()]), y.toArray(new Integer[y.size()]), colors.toArray(new Integer[colors.size()]));
        return board;
    }

    private static int scoreBoard(Board board) {
        int[] counters = {0, 0, 0, 0, 0, 0, 0, 0, 0};

        // Horizontal spans
        for (int y = 0; y < height; y++) {
            int score = board.slots[y][0] + board.slots[y][1] + board.slots[y][2];
            for (int x = 3; x < width; x++) {
                score += board.slots[y][x];
                counters[score + 4]++;
                score -= board.slots[y][x - 3];
            }
        }
        // Vertical spans
        for (int x = 0; x < width; x++) {
            int score = board.slots[0][x] + board.slots[1][x] + board.slots[2][x];
            for (int y = 3; y < height; y++) {
                score += board.slots[y][x];
                counters[score + 4]++;
                score -= board.slots[y - 3][x];
            }
        }
        // Down-right (and up-left) diagonals
        for (int y = 0; y < height - 3; y++) {
            for (int x = 0; x < width - 3; x++) {
                int score = 0;
                for (int idx = 0; idx < 4; idx++) {
                    score += board.slots[y+idx][x+idx];
                }
                counters[score + 4]++;
            }
        }
        // up-right (and down-left) diagonals
        for (int y = 3; y < height; y++) {
            for (int x = 0; x < width - 3; x++) {
                int score = 0;
                for (int idx = 0; idx < 4; idx++) {
                    score += board.slots[y-idx][x+idx];
                }
                counters[score + 4]++;
            }
        }

        if (counters[0] != 0) {
            return yellowWins;
        } else if (counters[8] != 0) {
            return orangeWins;
        } else {
            return counters[5] + 2 * counters[6] + 5 * counters[7] + 10 * counters[8]
                    - counters[3] - 2 * counters[2] - 5 * counters[1] - 10 * counters[0];
        }
    }

    private static ScoreMoveTuple abMinimax(boolean maximizeOrMinimize, int color, int depth, Board board) {

        if (0 == depth) {
            ScoreMoveTuple smt = new ScoreMoveTuple();
            smt.score = scoreBoard(board);
            smt.move = -1;
            return smt;
        } else {
            int bestScore = maximizeOrMinimize ? -10000000 : 10000000;
            int bestMove = -1;
            for (int column = 0; column < width; column++) {
                if (board.slots[0][column] != 0) {
                    continue;
                }
                int rowFilled = board.dropDisk(column, color);
                if (rowFilled == -1) {
                    continue;
                }
                int s = scoreBoard(board);
                if (s == (maximizeOrMinimize ? orangeWins : yellowWins)) {
                    bestMove = column;
                    bestScore = s;
                    board.slots[rowFilled][column] = 0;
                    break;
                }
                int moveInner, scoreInner;
                if (depth>1) {
                    ScoreMoveTuple inner = abMinimax(!maximizeOrMinimize, -color, depth - 1, board);
                    moveInner = inner.move;
                    scoreInner = inner.score;
                } else {
                    moveInner = -1;
                    scoreInner = s;
                }
                board.slots[rowFilled][column] = 0;
                if ((depth == g_maxDepth) && g_debug) {
                    System.out.printf("Depth %d, placing on %d, score:%d\n", depth, column, scoreInner);
                }
                if (maximizeOrMinimize) {
                    if (scoreInner >= bestScore) {
                        bestScore = scoreInner;
                        bestMove = column;
                    }
                } else {
                    if (scoreInner <= bestScore) {
                        bestScore = scoreInner;
                        bestMove = column;
                    }
                }
            }
            ScoreMoveTuple smt = new ScoreMoveTuple();
            smt.score = bestScore;
            smt.move = bestMove;
            return smt;
        }
    }

    static class ScoreMoveTuple {

        public int score;
        public int move;
    }
}
